General Instructions: 

Problem 1: Naive Bayes for Customer Churn Using 
Observed Data [50 pts] 
Background 
You are a data scientist working for a subscription-based company. Your goal is to predict 
whether a customer will churn (Y=1) or not churn (Y=0) using behavioral data. 
You will build two Naive Bayes models, each using a different feature set. For each model, you 
will: 

1.  Estimate probabilities from data 
2.  Apply Laplace smoothing 
3.  Compute posterior churn probabilities 
4.  Compare model conclusions 
Requirements 
●  Use Naive Bayes classification 
●  Derive all probabilities from the provided datasets 
●  Use Laplace (𝛂=1) smoothing for all conditional probabilities 
●  Compute normalized posterior probabilities 
●  Show all calculations clearly 
●  Use a calculator for computing, do not use machine learning libraries 
Given (common to both models) 

Let Y represent the event churn. From historical data, the overall churn rate is: 
●  P(Y = 1) = 0.3 
●  P(Y = 0) = 0.7 

Model A — Support-related features 
Feature definitions 

●  (X1 = 1): Customer opened many support tickets last month 
●  (X2 = 1): Customer spent a lot of time on help / FAQ pages 


Training dataset (historical sample) 
Churn (Y)  X₁ = 1  X₁ = 0  Total 
Y = 1       24      6       30 
Y = 0       14      56      70 
 
Churn (Y)  X₂ = 1  X₂ = 0  Total 
Y = 1       25      5       30 
Y = 0       13      57      70 

(Note: each feature table is estimated independently, as in Naive Bayes.) 
For an observed customer with: X1 = 1, X2 = 1, answer the following questions: 
Question A1 — Estimate probabilities 
Using Laplace smoothing, estimate: 
●  [5 pts] P(X1 = 1 | Y = 1), P(X1 = 1 | Y = 0) 
●  [5 pts] P(X2 = 1 | Y = 1), P(X2 = 1 | Y = 0)



In [7]:
a = 1  # Laplace smoothing parameter
v = 2  # Number of possible values for each feature (binary features)

p_x1_given_churn = (24+a) / (30+v*a)
p_x1_given_no_churn = (14+a) / (70+v*a)

p_x2_given_churn = (25+a) / (30+v*a)
p_x2_given_no_churn = (13+a) / (70+v*a)

print(f"P(X1 = 1 | Y = 1) = {p_x1_given_churn:.4f}")
print(f"P(X1 = 1 | Y = 0) = {p_x1_given_no_churn:.4f}") 
print(f"P(X2 = 1 | Y = 1) = {p_x2_given_churn:.4f}")
print(f"P(X2 = 1 | Y = 0) = {p_x2_given_no_churn:.4f}")


P(X1 = 1 | Y = 1) = 0.7812
P(X1 = 1 | Y = 0) = 0.2083
P(X2 = 1 | Y = 1) = 0.8125
P(X2 = 1 | Y = 0) = 0.1944



Question A2 — Posterior probabilities 
Using your smoothed estimates and Naive Bayes, compute the normalized posterior 
probabilities: 
●  [4 pts] P(Y = 1 | X1 = 1, X2 = 1) 
●  [4 pts] P(Y = 0 | X1 = 1, X2 = 1) 
 

In [8]:
# P(Y = 1 | X1 = 1, X2 = 1) = P(X1 = 1 | Y = 1) * P(X2 = 1 | Y = 1) * P(Y = 1) / P(X1 = 1, X2 = 1)

p_churn = 30 / 100  # P(Y = 1)
p_no_churn = 70 / 100  # P(Y = 0)

p_x1_x2_given_churn = p_x1_given_churn * p_x2_given_churn

p_x1_x2_given_no_churn = p_x1_given_no_churn * p_x2_given_no_churn
p_x1_x2 = p_x1_x2_given_churn * p_churn + p_x1_x2_given_no_churn * p_no_churn

p_y1_given_x1_x2 = (p_x1_x2_given_churn * p_churn) / p_x1_x2
p_y0_given_x1_x2 = (p_x1_x2_given_no_churn * p_no_churn) / p_x1_x2

print(f"P(Y = 1 | X1 = 1, X2 = 1) = {p_y1_given_x1_x2:.4f}")
print(f"P(Y = 0 | X1 = 1, X2 = 1) = {p_y0_given_x1_x2:.4f}")

P(Y = 1 | X1 = 1, X2 = 1) = 0.8704
P(Y = 0 | X1 = 1, X2 = 1) = 0.1296


Question A3 — Prediction 
Based on Model A: 

●  [1 pts] Does the model predict churn or no churn? 

    the model predicts churn because P(Y=1|X1=1,X2=1) > P(Y=0|X1=1,X2=1)

    
●  [1 pts] What is the predicted probability? 

    the predicted probability is P(Y=1|X1=1,X2=1) = 0.8704




Model B — Pricing and competition features 
Feature definitions 
●  (Z1 = 1): Customer changed pricing plan last month 
●  (Z2 = 1): Customer received a competitor’s marketing email 
Training dataset (historical sample) 
Churn (Y)  Z₁ = 1  Z₁ = 0  Total 
Y = 1  18  12  30 
Y = 0  21  49  70 
 
Churn (Y)  Z₂ = 1  Z₂ = 0  Total 
Y = 1  15  15  30 

Y = 0  28  42  70 
 
For an observed customer with: Z1 = 1, Z2 = 1, answer the following questions: 
Question B1 — Estimate probabilities 
Using Laplace smoothing, estimate: 
●  [5 pts] P(Z1 = 1 | Y = 1), P(Z1 = 1 | Y = 0) 
●  [5 pts] P(Z2 = 1 | Y = 1), P(Z2 = 1 | Y = 0) 


In [ ]:
a = 1  # Laplace smoothing parameter
v = 2  # Number of possible values for each feature (binary features)

# Z1 conditional probabilities with Laplace smoothing
p_z1_given_churn = (18 + a) / (30 + v * a)
p_z1_given_no_churn = (21 + a) / (70 + v * a)

# Z2 conditional probabilities with Laplace smoothing
p_z2_given_churn = (15 + a) / (30 + v * a)
p_z2_given_no_churn = (28 + a) / (70 + v * a)

print(f"P(Z1 = 1 | Y = 1) = (18+1)/(30+2) = 19/32 = {p_z1_given_churn:.4f}")
print(f"P(Z1 = 1 | Y = 0) = (21+1)/(70+2) = 22/72 = {p_z1_given_no_churn:.4f}")
print(f"P(Z2 = 1 | Y = 1) = (15+1)/(30+2) = 16/32 = {p_z2_given_churn:.4f}")
print(f"P(Z2 = 1 | Y = 0) = (28+1)/(70+2) = 29/72 = {p_z2_given_no_churn:.4f}")

Question B2 — Posterior probabilities 
Using your smoothed estimates and Naive Bayes, compute the normalized posterior 
probabilities: 
●  [4 pts] P(Y = 1 | Z1 = 1, Z2 = 1) 
●  [4 pts] P(Y = 0 | Z1 = 1, Z2 = 1) 


In [ ]:
# Prior probabilities
p_churn = 0.3
p_no_churn = 0.7

# Unnormalized posteriors (joint likelihoods)
joint_churn = p_z1_given_churn * p_z2_given_churn * p_churn
joint_no_churn = p_z1_given_no_churn * p_z2_given_no_churn * p_no_churn

print(f"Unnormalized P(Y=1) term: {p_z1_given_churn:.4f} * {p_z2_given_churn:.4f} * {p_churn} = {joint_churn:.6f}")
print(f"Unnormalized P(Y=0) term: {p_z1_given_no_churn:.4f} * {p_z2_given_no_churn:.4f} * {p_no_churn} = {joint_no_churn:.6f}")

# Normalize
total = joint_churn + joint_no_churn
p_y1_given_z1_z2 = joint_churn / total
p_y0_given_z1_z2 = joint_no_churn / total

print(f"\nP(Y = 1 | Z1 = 1, Z2 = 1) = {joint_churn:.6f} / {total:.6f} = {p_y1_given_z1_z2:.4f}")
print(f"P(Y = 0 | Z1 = 1, Z2 = 1) = {joint_no_churn:.6f} / {total:.6f} = {p_y0_given_z1_z2:.4f}")

Question B3 — Prediction 
Based on Model B: 
●  [1 pts] Does the model predict churn or no churn? 
●  [1 pts] What is the predicted probability? 
Model comparison and interpretation 
Compare the predictions from Model A and Model B: 
●  [5 pts] Why are the predicted churn probabilities so different? 
●  [5 pts] Which model’s prediction is more trustworthy? Why? 
 

**Question B3 — Prediction**

- The model predicts **churn** because P(Y=1|Z1=1,Z2=1) ≈ 0.5083 > P(Y=0|Z1=1,Z2=1) ≈ 0.4917
- The predicted probability is **P(Y=1|Z1=1,Z2=1) ≈ 0.5083**

---

**Model Comparison and Interpretation**

**Why are the predicted churn probabilities so different?**

Model A predicts ~87% churn while Model B predicts only ~51% churn. The difference comes from the discriminative power of the features. Model A's features (support tickets, FAQ page usage) show a dramatic separation between churners and non-churners — for example, P(X1=1|Y=1) ≈ 0.78 vs P(X1=1|Y=0) ≈ 0.21, a likelihood ratio of ~3.75. Model B's features (pricing plan changes, competitor emails) are distributed much more evenly across both classes — for example, P(Z1=1|Y=1) ≈ 0.59 vs P(Z1=1|Y=0) ≈ 0.31, a ratio of only ~1.94, and Z2 is even weaker with a ratio of ~1.24. When features provide only marginally different likelihoods between classes, the posterior barely moves away from the prior (0.3), whereas highly discriminative features push the posterior much more strongly.

**Which model's prediction is more trustworthy? Why?**

Model A is more trustworthy. Its features (support ticket volume and FAQ page time) are strongly and directly linked to customer dissatisfaction — a dissatisfied customer seeking help frequently is exhibiting clear churn-indicative behavior. The conditional probabilities show large separations between classes, meaning these features carry strong predictive signal. Model B's features are inherently noisier: a customer may change pricing plans for cost optimization without intending to leave, and receiving a competitor's email doesn't mean they acted on it. The near-50/50 posterior in Model B indicates the features barely distinguish churners from non-churners, making its prediction unreliable — essentially no better than a coin flip.

Problem 2 [50 pts]:  
Complete the Forward-Backtracking process for the Hidden Markov Model (HMM) following the 
table format presented in the lecture slides. The HMM structure and observation sequence are 
exactly the same as in the lecture example. The only changes are the transition and emission 
probabilities, which are given below. Use these probability tables to perform your calculations. 

Transition: 
From → To  Sunny  Cloudy  Rainy 
Sunny  0.33  0.67  0.00 
Cloudy  0.33  0.00  0.67 
Rainy  0.33  0.33  0.33 
 
Emission: 
Weather → Behavior  Walk  Umbrella 
Sunny  4/4 = 1.0  0/3 = 0.0 
Cloudy  2/3 ≈ 0.67  1/3 ≈ 0.33 
Rainy  1/3 ≈ 0.33  2/3 ≈ 0.67 
You can manually compute (show the steps), code (submit the code through Github link) or use 
a spreadsheet (submit the spreadsheet) to work out the solutions.

In [ ]:
# ============================================================
# Problem 2: HMM Forward-Backward Algorithm
# ============================================================
# States: Sunny (S), Cloudy (C), Rainy (R)
# Observations: Walk (W), Umbrella (U)
# Observation sequence: [Walk, Umbrella, Walk]
# Initial probabilities: uniform (1/3 each)

states = ["Sunny", "Cloudy", "Rainy"]
obs_names = ["Walk", "Umbrella"]
observations = [0, 1, 0]  # 0=Walk, 1=Umbrella
T = len(observations)

# Initial state distribution (uniform)
pi = [1/3, 1/3, 1/3]

# Transition matrix A[i][j] = P(state_j | state_i)
A = [
    [0.33, 0.67, 0.00],  # From Sunny
    [0.33, 0.00, 0.67],  # From Cloudy
    [0.33, 0.33, 0.33],  # From Rainy
]

# Emission matrix B[i][j] = P(obs_j | state_i)
B = [
    [1.0,  0.0 ],  # Sunny:  Walk=1.0,   Umbrella=0.0
    [2/3,  1/3 ],  # Cloudy: Walk=2/3,   Umbrella=1/3
    [1/3,  2/3 ],  # Rainy:  Walk=1/3,   Umbrella=2/3
]

N = len(states)

# ============================================================
# FORWARD PASS: alpha[t][i] = P(o_1,...,o_t, q_t = S_i)
# ============================================================
alpha = [[0.0] * N for _ in range(T)]

# Initialization (t=0)
for i in range(N):
    alpha[0][i] = pi[i] * B[i][observations[0]]

# Induction (t=1..T-1)
for t in range(1, T):
    for j in range(N):
        alpha[t][j] = sum(alpha[t-1][i] * A[i][j] for i in range(N)) * B[j][observations[t]]

print("=" * 60)
print("FORWARD PASS (alpha)")
print("=" * 60)
for t in range(T):
    obs_label = obs_names[observations[t]]
    print(f"\nt={t+1} (Observation: {obs_label})")
    for i in range(N):
        print(f"  alpha_{t+1}({states[i]:6s}) = {alpha[t][i]:.6f}")

p_obs_forward = sum(alpha[T-1])
print(f"\nP(O) from forward pass = {p_obs_forward:.6f}")

# ============================================================
# BACKWARD PASS: beta[t][i] = P(o_{t+1},...,o_T | q_t = S_i)
# ============================================================
beta = [[0.0] * N for _ in range(T)]

# Initialization (t=T-1)
for i in range(N):
    beta[T-1][i] = 1.0

# Induction (t=T-2..0)
for t in range(T-2, -1, -1):
    for i in range(N):
        beta[t][i] = sum(A[i][j] * B[j][observations[t+1]] * beta[t+1][j] for j in range(N))

print("\n" + "=" * 60)
print("BACKWARD PASS (beta)")
print("=" * 60)
for t in range(T):
    obs_label = obs_names[observations[t]]
    print(f"\nt={t+1} (Observation: {obs_label})")
    for i in range(N):
        print(f"  beta_{t+1}({states[i]:6s}) = {beta[t][i]:.6f}")

# Verify: P(O) from backward pass
p_obs_backward = sum(pi[i] * B[i][observations[0]] * beta[0][i] for i in range(N))
print(f"\nP(O) from backward pass = {p_obs_backward:.6f}")

# ============================================================
# POSTERIOR (SMOOTHED) PROBABILITIES: gamma[t][i] = P(q_t = S_i | O)
# gamma_t(i) = alpha_t(i) * beta_t(i) / P(O)
# ============================================================
print("\n" + "=" * 60)
print("POSTERIOR PROBABILITIES (gamma = alpha * beta / P(O))")
print("=" * 60)

gamma = [[0.0] * N for _ in range(T)]
for t in range(T):
    obs_label = obs_names[observations[t]]
    print(f"\nt={t+1} (Observation: {obs_label})")
    for i in range(N):
        gamma[t][i] = alpha[t][i] * beta[t][i] / p_obs_forward
        print(f"  P({states[i]:6s} | O) = {alpha[t][i]:.6f} * {beta[t][i]:.6f} / {p_obs_forward:.6f} = {gamma[t][i]:.4f}")
    most_likely = states[gamma[t].index(max(gamma[t]))]
    print(f"  -> Most likely state: {most_likely}")

# ============================================================
# SUMMARY TABLE
# ============================================================
print("\n" + "=" * 60)
print("SUMMARY TABLE")
print("=" * 60)
print(f"{'t':>3} | {'Obs':>9} | {'alpha_S':>9} {'alpha_C':>9} {'alpha_R':>9} | {'beta_S':>9} {'beta_C':>9} {'beta_R':>9} | {'gamma_S':>8} {'gamma_C':>8} {'gamma_R':>8} | {'Best':>6}")
print("-" * 120)
for t in range(T):
    obs_label = obs_names[observations[t]]
    best = states[gamma[t].index(max(gamma[t]))]
    print(f"{t+1:>3} | {obs_label:>9} | {alpha[t][0]:>9.6f} {alpha[t][1]:>9.6f} {alpha[t][2]:>9.6f} | {beta[t][0]:>9.6f} {beta[t][1]:>9.6f} {beta[t][2]:>9.6f} | {gamma[t][0]:>8.4f} {gamma[t][1]:>8.4f} {gamma[t][2]:>8.4f} | {best:>6}")

print(f"\nMost likely state sequence: {' -> '.join(states[gamma[t].index(max(gamma[t]))] for t in range(T))}")